torchmultimodal/models/gpt.py

```diff
@@ -28,12 +28,221 @@ class TransformerLayerOutput(NamedTuple):
+            positions to attend to and 0s for masked positions. Defaults to ``None``.
+        head_mask (Tensor, optional): Tensor of dimension ``(b, h, total_seq_len, total_seq_len)``. Contains 1s
+            for attention heads to use and 0s for masked heads. Defaults to ``None``.
+        use_cache (bool, optional): If ``True``, caches past key/value tensors for faster decoding. If ``False``,
+            recomputes key and value for each decoding step. Defaults to ``False``.
+        causal (bool. optional): If ``True``, use causal attention. Defaults to ``False``.
+        right_shift (bool): If ``True``, shifts the embedding vectors to the right and prepends it with start of
+            sentence token. Defaults to ``False``. This option is disregarded during training mode
+        return_attn_weights (bool, optional): If ``True``, returns attention probabilities of each transformer
+            layer. Defaults to ``False``.
+        return_hidden_states (bool): If ``True``, returns the embeddings of each transformer layer. Defaults to
+            ``False``.
+
+    Raises:
+        AttributeError: If input tokenizer does not implement methods ``encode`` and ``decode`` or output
+            tokenizer does not implement methods ``encode``, ``decode`` and ``lookup``.
+    """
+
+    def __init__(
+        self,
+        d_model: int,
+        num_tokens: int,
+        in_tokenizer: nn.Module,
+        out_tokenizer: nn.Module,
+        mm_decoder: nn.Module,
+    ) -> None:
+        super().__init__()
+        if not all(
+            [hasattr(in_tokenizer, attr_name) for attr_name in ["encode", "decode"]]
+        ):
+            raise AttributeError(
+                "Input modality tokenizer must have methods 'encode', 'decode' and 'lookup'."
+            )
+
+        if not all(
+            [
+                hasattr(out_tokenizer, attr_name)
+                for attr_name in ["encode", "decode", "lookup"]
+            ]
+        ):
+            raise AttributeError(
+                "Output modality tokenizer must have methods 'encode', 'decode' and 'lookup'."
+            )
+
+        self.d_model = d_model
+        self.num_tokens = num_tokens
+        self.in_tokenizer = in_tokenizer
+        self.out_tokenizer = out_tokenizer
+        self.mm_decoder = mm_decoder
+        self.norm = nn.LayerNorm(d_model)
+        self.to_logit = nn.Linear(d_model, num_tokens, bias=False)
+        # This will give us equal probabilities after the soft max layer initially to avoid biasing
+        # towards any particular prediction category
+        self.to_logit.weight.data.copy_(torch.zeros(self.num_tokens, self.d_model))
+
+    def forward(
+        self,
+        in_modality: Optional[Tensor] = None,
+        out_modality: Optional[Tensor] = None,
+        in_pos_ids: Optional[Tensor] = None,
+        out_pos_ids: Optional[Tensor] = None,
+        attn_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
```

Comment:
I know we've been consistent with the name `head_mask`, but it may be confusing here since you have an actual head. maybe consider renaming to `attn_head_mask` or similar? your call

Reply:
Not sure if I understand the proposed confusion here. 
`head_mask` is just padding mask for each attention head. 
For now we can just be consistent with the rest of the codebase: https://github.com/facebookresearch/multimodal/search?q=head_mask

Reply:
This `head_mask` is in reference to the attention head but you also have the logits head, so I was just wondering if we should clarify that this is for the attention head and not the logits head to avoid confusion

Reply:
I see. you meant `head` could be confused with `head` for the logit prediction.